Orm/Xtensive.Orm/Orm/Model/IndexInfo.cs

```diff
@@ -428,16 +395,34 @@ public IndexInfo(TypeInfo declaringType, IndexAttributes indexAttributes)
     public IndexInfo(TypeInfo reflectedType, IndexAttributes indexAttributes, IndexInfo ancestorIndex)
       : this()
     {
-      declaringType = ancestorIndex.DeclaringType;
-      this.reflectedType = reflectedType;
+      DeclaringType = ancestorIndex.DeclaringType;
+      ReflectedType = reflectedType;
       attributes = indexAttributes;
 
       fillFactor = ancestorIndex.FillFactor;
       filterExpression = ancestorIndex.FilterExpression;
-      declaringIndex = ancestorIndex.DeclaringIndex;
+      DeclaringIndex = ancestorIndex.DeclaringIndex;
       shortName = ancestorIndex.ShortName;
     }
 
+    /// <summary>
+    /// Used for cloning only
+    /// </summary>
+    private IndexInfo(IndexInfo original)
+    {
+      shortName = original.shortName;
+      Name = original.Name;
+      KeyColumns = original.KeyColumns;
+      IncludedColumns = original.IncludedColumns;
+      ValueColumns = original.ValueColumns;
+      ReflectedType = original.ReflectedType;
+      attributes = original.attributes;
+      DeclaringType = original.DeclaringIndex.DeclaringType;
+      fillFactor = original.DeclaringIndex.FillFactor;
+      filterExpression = original.DeclaringIndex.FilterExpression;
+      DeclaringIndex = original.DeclaringIndex.DeclaringIndex;
+    }
```

Comment:
I see another violation of this rule above yours constructor. Since you shed light on this code region, it should also be fixed, I'll fix it after merge.

Reply:
> Also, let's make no exceptions and use comma at the end of the sentence in summary.

dot, not comma?

Reply:
> Members should be arranged by access level from most accessible to least accessible, from public to private.

But public properties follow private fields.
What is the order of initialization.
I think it is better to init in the order of declaration

Reply:
> But public properties follow private fields.
> What is the order of initialization.
> I think it is better to init in the order of declaration.

First, we group members by type - fields, properties, static methods, instance methods, constructors, finalizers. General rule within each group is to have members in order from most accessible to least accessible as I said. 

In this case private constructor is less usable and has basically one job and probably will be the least readable ```ctor``` among all, so it should be at the end of constructors list.

This is preferable structure, but, of course, there are some exceptions, in particular for methods and constructors, in favor of reading we do allow members to be places against general rule, for example:
- methods that implement interface are likely to be in one group, sometimes they are in ```#region```s;
- methods with the same name may also be grouped no matter what access modifier they have, but it is better to apply general rule within the methods, for instance ```public void Add()```, then ```internal void Add()``` and then ```private void Add()```/```private void AddInternal()```.
- various helper methods, they mix and match instance and static methods but most of the time they are private and small; with proper name they don't require going into them to understand an algorithm which uses them. Since they are less important in general, they tend to be moved to the end of methods region.
- serialization/deserialization members (methods and ctors,); they are read extremely rarely so they are preferred to be after ctors.
- GetEnumerator  methods are also very simple and not very important to be read, so they are preferred to be moved closer to the end of methods region.

These are exceptions I remembered but probably there are some others. If I saw something I just point to it, when I have time to properly explain why I try to, like now.

As you may already understood, the main idea of such structure is to bring up the members of their group (or whole class) which are more important when reading or scanning through the class and to push down the bottom of the class the members which are less likely to be read. We chose this structure and I try to maintain it because I see its effectiveness. After year of being junior in the project I became the only guy who supports it and had to read unseen code a lot and this structure, in my humble opinion, helped me to not waste some time skipping unimportant members like constructors, helpers, serialization members, etc.

Reply:
On the other side, when you participate in 10+ different projects, and avery project has its own rules, it is hardly possible to remember compex rulesets, anything behind "fields, properties, constructors"

So I'm always trying to guess the order by example and forget them on next day

Reply:
I understand, It's ok that you can't remember it. More important that you apply changes that requested.

Reply:
Best way to enforce code style rules - to develop a Roslyn Analyzer.